File: src/main/java/org/maxur/ddd/Launcher.java
package org.maxur.ddd;


import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.yaml.snakeyaml.Yaml;

import javax.validation.ConstraintViolation;
import javax.validation.Validation;
import javax.validation.Validator;
import javax.validation.ValidatorFactory;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.function.Function;

import static io.baratine.web.Web.bean;
import static io.baratine.web.Web.go;
import static io.baratine.web.Web.include;
import static io.baratine.web.Web.property;
import static io.baratine.web.Web.service;
import static java.lang.String.format;
import static java.util.stream.Collectors.joining;

/**
 * Application Launcher.
 *
 * @author myunusov
 * @version 1.0
 * @since <pre>10.07.2016</pre>
 */
@Slf4j
public final class Launcher {

    private static final String CONFIG_YAML = "./conf/config.yaml";

    private Launcher() {
    }

    /**
     * The entry point of application.
     *
     * @param args the input arguments
     */
    public static void main(String[] args) {
        rest(args);
    }

    private static void rest(String[] args) {
        final ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory();
        bean(validatorFactory).to(ValidatorFactory.class);
        final UserConfig config = config(validatorFactory.getValidator());
        bean(config).to(UserConfig.class);
        property("server.port", String.valueOf(config.getRest().getPort()));
        include(SysConfig.class);
        service(TaskService.class);
        include(TaskResource.class);
        log.info(format("Start Service on http://%s", config.getRest()));
        go(args);
    }

    private static UserConfig config(final Validator validator) {
        final Yaml yaml = new Yaml();
        final UserConfig config;
        try (InputStream in = Files.newInputStream(Paths.get(CONFIG_YAML))) {
            config = yaml.loadAs(in, UserConfig.class);
            val constraintViolation = validator.validate(config);
            if (!constraintViolation.isEmpty()) {
                throw new IllegalArgumentException(
                    format(
                        "User configuration error:\n %s",
                        constraintViolation.stream()
                            .map(validationMessageFrom())
                            .collect(joining(";\n"))
                    )
                );
            }
        } catch (IOException e) {
            log.error("Config file '" + CONFIG_YAML +"' is not accessible");
            throw new IllegalArgumentException(e);
        }
        return config;
    }

    private static Function<ConstraintViolation<UserConfig>, String> validationMessageFrom() {
        return cv -> format(
            "Field '%s.%s' %s",
            cv.getLeafBean().getClass().getSimpleName(),
            cv.getPropertyPath(),
            cv.getMessage()
        );
    }


}

File: src/main/kotlin/org/maxur/ddd/Launcher.kt
package org.maxur.ddd

import org.maxur.ddd.params.ConfigParams
import org.maxur.mserv.core.builder.Kotlin
import org.maxur.mserv.core.builder.hocon
import org.maxur.mserv.core.domain.BaseService
import org.maxur.mserv.core.embedded.WebServer
import org.maxur.mserv.core.service.properties.Properties
import org.slf4j.LoggerFactory

/**
 * Application Launcher
 *
 * @author myunusov
 * @version 1.0
 * @since <pre>12.06.2017</pre>
 */
object Launcher {

    private fun log() = LoggerFactory.getLogger(Launcher::class.java)

    /**
     * Command line entry point. This method kicks off the building of a application  object
     * and executes it.
     *
     * @param args - arguments of command.
     */
    @JvmStatic
    fun main(args: Array<String>) {
        Kotlin.service {
            name = ":name"
            packages += "org.maxur.ddd"
            properties += hocon()
            services += rest {
                afterStart += this@Launcher::afterWebServiceStart
            }
            afterStart += this@Launcher::afterStart
            beforeStop += { _ -> log().info("Microservice is stopped") }
            onError += { exception -> log().error(exception.message, exception) }
        }.start()
    }

    fun afterStart(configParams: ConfigParams, config: Properties, service: BaseService) {
        log().info("Properties Source is '${config.sources.get(0).format}'\n")
        configParams.log()
        log().info("${service.name} is started")
    }

    fun afterWebServiceStart(service: WebServer) {
        log().info("${service.name} is started on ${service.baseUri}\"")
        log().info(service.entries().toString())
    }

}

